src/stream/src/executor/source/source_executor.rs

```diff
@@ -132,96 +133,67 @@ impl<S: StateStore> SourceExecutor<S> {
     }
 }
 
+type SourceReaderMessage =
+    Either<StreamExecutorResult<Barrier>, StreamExecutorResult<StreamChunkWithState>>;
+type SourceReaderArm = BoxStream<'static, SourceReaderMessage>;
+type SourceReaderStream =
+    SelectWithStrategy<SourceReaderArm, SourceReaderArm, fn(&mut ()) -> PollNext, ()>;
+
 struct SourceReader {
-    /// The reader for stream source.
-    stream_reader: Arc<Mutex<Option<Box<SourceStreamReaderImpl>>>>,
     /// The reader for barrier.
     barrier_receiver: UnboundedReceiver<Barrier>,
-    /// Expected barrier latency in ms. If there are no barrier within the expected barrier
-    /// latency, source will stall.
-    expected_barrier_latency_ms: u64,
+
+    /// The reader for stream source.
+    source_chunk_reader: Box<SourceStreamReaderImpl>,
 }
 
 impl SourceReader {
-    #[try_stream(ok = StreamChunkWithState, error = RwError)]
-    async fn stream_reader(
-        stream_reader: Arc<Mutex<Option<Box<SourceStreamReaderImpl>>>>,
-        #[expect(unused_variables)] notifier: Arc<Notify>,
-        #[expect(unused_variables)] expected_barrier_latency_ms: u64,
-        mut abort_notifier: UnboundedReceiver<()>,
-    ) {
-        'outer: loop {
-            let mut reader = stream_reader.lock().await.take().unwrap();
-            let chunk_result: Option<Result<StreamChunkWithState>>;
-
-            {
-                let chunk_future = reader.next();
-                let abort_future = abort_notifier.recv();
-
-                pin_mut!(chunk_future);
-                pin_mut!(abort_future);
+    /// Receive barriers from barrier manager with the channel, error on channel close.
+    #[try_stream(ok = Barrier, error = StreamExecutorError)]
+    async fn barrier_receiver(mut rx: UnboundedReceiver<Barrier>) {
+        while let Some(barrier) = rx.recv().await {
+            yield barrier;
+        }
+        bail!("barrier reader closed unexpectedly");
```

Comment:
may also pending here. Not sure whether replacing a terminated stream would cause some confusing result...

Reply:
or panic here?